### ISSUE TITLE
Batch context manager fails when exiting

### ISSUE THREAD

**Issue by Raab70** [googleapis/python-storage#86]
Thanks for stopping by to let us know something could be better!

#### Environment details

  - OS type and version: OSX 10.15
  - Python version: `python --version` 3.6.8
  - pip version: `pip --version` 18.1
  - `google-cloud-storage` version: `pip show google-cloud-storage`
```
$ pip show google-cloud-storage
Name: google-cloud-storage
Version: 1.26.0
Summary: Google Cloud Storage API client library
Home-page: https://github.com/googleapis/python-storage
Author: Google LLC
Author-email: googleapis-packages@google.com
License: Apache 2.0
Location: /Users/rharrigan/.virtualenvs/chart/lib/python3.6/site-packages
Requires: google-auth, google-cloud-core, google-resumable-media
Required-by: 
```

#### Steps to reproduce

  1. Create a batch with a multiple of 100 objects
  2. ValueError is raised on exit

#### Code example

```python
storage_client = get_storage_client()
bucket = storage_client.bucket(BUCKET)
with storage_client.batch():
    for idx in range(100):
        blob = bucket.blob(f"test/{idx}")
        blob.upload_from_string("test")

```

#### Stack trace
```
Traceback (most recent call last):
  File "main.py", line 75, in <module>
    main()
  File "main.py", line 68, in main
    blob.upload_from_string("test")
  File "/.virtualenvs/chart/lib/python3.6/site-packages/google/cloud/storage/batch.py", line 292, in __exit__
    self.finish()
  File "/.virtualenvs/chart/lib/python3.6/site-packages/google/cloud/storage/batch.py", line 267, in finish
    headers, body, timeout = self._prepare_batch_request()
  File "/.virtualenvs/chart/lib/python3.6/site-packages/google/cloud/storage/batch.py", line 209, in _prepare_batch_request
    raise ValueError("No deferred requests")
ValueError: No deferred requests
```

It seems like the context manager is not handling the case of exiting gracefully if all items have been sent.

**Comment by egorklimov** [comment]
Hello! @Raab70 

I faced the same error, why did you close the issue?

**Comment by dahburj** [comment]
> ?

Same here

**Comment by dahburj** [comment]
@Raab70  Same here 🙋🏻‍♂️

**Comment by zchauvin** [comment]
For anyone finding this, my problem when getting this error was that my code in the `with` block was no-oping, meaning there were no requests for the GCS batch to process. Here's a minimal example: 

```
... create storage_client
with storage_client.batch():
    print('hello, world!')
```

In this case, the batch has no requests, and throws a ValueError on exit, so you should check that there are in fact requests in the batch if you're getting this error.

**Comment by Mikkel84** [comment]
storage_client.batch() only works for update of metadata and deletion of blobs. It does not work for uploads/downloads.

https://cloud.google.com/storage/docs/batch#code-samples-batch-requests
